<a href="https://colab.research.google.com/github/ProgramasParaFisicaDeParticulas/blindsampleanalysiswithcsv7-wigilm/blob/main/Z0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go


data=pd.read_csv('https://raw.githubusercontent.com/ProgramasParaFisicaDeParticulas/blindsampleanalysiswithcsv7-wigilm/main/Sample7.csv')
print(len(data))
data.head(10)

10000


Run      Event      pt1    eta1    phi1  Q1 type1  sigmaEtaEta1  \
0  163286  109060857  37.5667  2.2892  2.0526  -1    EE        0.0251   
1  163286  109275715  36.2901 -0.8373 -1.5859   1    EB        0.0078   
2  163286  109075352  25.9705 -0.6974  1.6360  -1    EB        0.0097   
3  163286  109169766  41.0075  1.4619 -0.5325   1    EB        0.0088   
4  163286  108947653  39.8985 -0.5927 -2.3947   1    EB        0.0153   
5  163286  109260503  41.9583 -0.4056 -2.7564   1    EB        0.0126   
6  163286  109491381  62.3848  1.0751 -1.7610  -1    EB        0.0083   
7  163286  109508443  25.1316 -2.1099 -1.8266   1    EE        0.0249   
8  163286  110268001  32.3642 -0.5340 -2.6718  -1    EB        0.0109   
9  163286  110085994  31.4123 -0.2164  1.2954  -1    EB        0.0118   

   HoverE1  isoTrack1  ...      pt2    eta2    phi2  Q2  type2  sigmaEtaEta2  \
0   0.0090     0.0000  ...  45.4315  1.4706 -1.1630   1     EB        0.0008   
1   0.0438     0.0000  ...  60.5754 -0.4896  1.0496  -1     EB        0.0112   
2   0.0407     6.2870  ...  45.2954 -2.0401  3.1187   1     EE        0.0260   
3   0.0000     0.0000  ...  45.9013  1.1561  2.4786  -1     EB        0.0086   
4   0.0000     2.5435  ...  34.8931 -2.2444  0.6106  -1     EE        0.0290   
5   0.0164    29.7964  ...  36.7997 -2.3813  0.0790   1     EE        0.0315   
6   0.0000     0.0000  ...  43.4772  1.5828  1.4717   1     EE        0.0108   
7   0.0000     0.0000  ...  38.0420  0.3717 -0.8307   1     EB        0.0127   
8   0.0843     4.7355  ...  40.6404  1.3910  0.0744   1     EB        0.0180   
9   0.1096    18.8325  ...  42.4105 -1.2758 -2.1235  -1     EB        0.0087   

  HoverE2  isoTrack2  isoEcal2  isoHcal2  
0  0.0000     0.0000    1.0190    0.0000  
1  0.0000     0.7185    1.8461    0.0000  
2  0.0280    15.2170    4.5337    3.8370  
3  0.0000     0.0000    2.4388    0.5676  
4  0.0000    12.4229    0.4534    0.9096  
5  0.0289     0.0000    4.9119    5.5508  
6  0.0195    15.3751    2.8608    3.1914  
7  0.1062     8.5834    2.9123    0.0000  
8  0.0000    12.4254    4.1619    7.4554  
9  0.0000    20.4377    6.4224    2.4006  

[10 rows x 22 columns]

Veamos como se comportan los detectores viendo las graficas de los ángulos phi y eta

In [3]:
fig_test=px.histogram(data.phi1)
fig_test.show()
fig_test2=px.histogram(data.phi2)
fig_test2.show()
fig_test3=px.scatter(x=data.phi1,y=data.eta1)
fig_test3.show()
fig_test4=px.scatter(x=data.phi2,y=data.eta2)
fig_test4.show()

In [4]:
def minvariant(pt1,pt2,eta1,eta2,phi1,phi2):
  return np.sqrt(2*pt1*pt2*(np.cosh(eta2-eta1)-np.cos(phi1-phi2)))
minv=minvariant(data['pt1'],data['pt2'],data['eta1'],data['eta2'],data['phi1'],data['phi2'])

fig_1=px.histogram(minv)
fig_1.show()

fig_2=px.histogram(data['Q1']+data['Q2'])
fig_2.show()

La masa invariante de los datos crudos recuerda a aquella del Z0. Parece que nuevamente los eventos están entre 60Gev/c^2 y 120 Gev/c^2. Sin embargo, ahora hay muchos eventos con carga total distinta de 0. Veamos que pasa si se quitan

In [5]:
data1=data[data.Q1+data.Q2==0]
minv1=minvariant(data1['pt1'],data1['pt2'],data1['eta1'],data1['eta2'],data1['phi1'],data1['phi2'])
print(len(data1))

fig_3=go.Figure()
fig_3.add_trace(go.Histogram(x=minv, name='eventos raw',nbinsx=200))
fig_3.add_trace(go.Histogram(x=minv1, name='eventos Q=0',nbinsx=200))
fig_3.update_layout(barmode='overlay')
fig_3.show()

7591


Vemos que ahora solamente quedan 7591 datos de los 10,000 con los que empezamos. Además, la gráfica se ha aplanado más del pico, se esperaría que el pico fuera más pronunciado después de haber eliminado estos datos. Revisemos si son electrones o muones los datos para ver si se encuentra algo interesante.


In [6]:
data_electron=data[((data.type1=='EE') | (data.type1=='EB')) & ((data.type2=='EE') | (data.type2=='EB'))]
print(len(data_electron))

10000


¡Todos los datos son electrones! Veamos como jugar con el parámetro sigmaetaeta para discriminar posibles jets que no corresponden a electrones.

> Bloc con sangría



In [7]:
tol=0.02
bins=1000
data_jets=data1[(data1.sigmaEtaEta1 <=tol )  & (data1.sigmaEtaEta2 <=tol)]
print(len(data_jets))
minvjets=minvariant(data_jets['pt1'],data_jets['pt2'],data_jets['eta1'],data_jets['eta2'],data_jets['phi1'],data_jets['phi2'])

fig_4=go.Figure()
fig_4.add_trace(go.Histogram(x=minv, name='eventos raw'))
fig_4.add_trace(go.Histogram(x=minv1, name='eventos Q=0'))
fig_4.add_trace(go.Histogram(x=minvjets, name='eventos Q=0 y sigmaetaeta<=tol'))
fig_4.update_layout(barmode='overlay')
fig_4.show()

5819


interesante como cambia la grafica de 15.94 a 15.95

In [8]:
tol_iso=15.945
data_iso=data_jets[(data_jets.isoEcal1+data_jets.isoHcal1+data_jets.isoTrack1<=tol_iso) & (data_jets.isoEcal2+data_jets.isoHcal2+data_jets.isoTrack2<=tol_iso)]
print(len(data_iso))
minviso=minvariant(data_iso['pt1'],data_iso['pt2'],data_iso['eta1'],data_iso['eta2'],data_iso['phi1'],data_iso['phi2'])
fig_4=go.Figure()
fig_4.add_trace(go.Histogram(x=minv, name='eventos raw',nbinsx=200))
fig_4.add_trace(go.Histogram(x=minv1, name='eventos Q=0',nbinsx=200))
fig_4.add_trace(go.Histogram(x=minvjets, name='sigmaetaeta<=tol',nbinsx=200))
fig_4.add_trace(go.Histogram(x=minviso, name='filtro iso',nbinsx=200))
fig_4.update_layout(barmode='overlay')
fig_4.show()

4453


Ahora vamos a ver la calidad de los electrones discriminando los datos que solo se detectan en un detector, empezemos con los datos originales y pongamosle un filtro de detector EE

In [9]:
electron_EE=data[((data.type1=='EE') & (data.type2=='EE'))]
print(len(electron_EE))
electron_EB=data[(data.type1=='EB') & (data.type2=='EB')]
print(len(electron_EB))
electron_mix=data[((data.type1=='EB') & (data.type2=='EE')) | ((data.type1=='EE') & (data.type2=='EB')) ]
print(len(electron_mix))
m_EE=minvariant(electron_EE['pt1'],electron_EE['pt2'],electron_EE['eta1'],electron_EE['eta2'],electron_EE['phi1'],electron_EE['phi2'])
m_EB=minvariant(electron_EB['pt1'],electron_EB['pt2'],electron_EB['eta1'],electron_EB['eta2'],electron_EB['phi1'],electron_EB['phi2'])
m_mix=minvariant(electron_mix['pt1'],electron_mix['pt2'],electron_mix['eta1'],electron_mix['eta2'],electron_mix['phi1'],electron_mix['phi2'])
fig_5=go.Figure()
fig_5.add_trace(go.Histogram(x=minv, name='RAW', nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_EB, name='BARREL', nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_mix, name='MIXTO', nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_EE, name='ENDCAP', nbinsx=200))
fig_5.update_layout(barmode='overlay')
fig_5.show()

1272
4981
3747


Los datos relevantes se conocen tienen el filtro Q1+Q2=0, se hará esto para todos los datos.

In [10]:
electron_EE1=electron_EE[electron_EE.Q1+electron_EE.Q2==0]
print("la longitud de endcap q=0 es " ,len(electron_EE1))
electron_EB1=electron_EB[electron_EB.Q1+electron_EB.Q2==0]
print("la longitud de barrel q=0 es ", len(electron_EB1))
electron_mix1=electron_mix[electron_mix.Q1+electron_mix.Q2==0]
print("la longitud de mix q=0 es ",len(electron_mix1))
print("la longitud de raw q=0 es ",len(data1))
m_EE1=minvariant(electron_EE1['pt1'],electron_EE1['pt2'],electron_EE1['eta1'],electron_EE1['eta2'],electron_EE1['phi1'],electron_EE1['phi2'])
m_EB1=minvariant(electron_EB1['pt1'],electron_EB1['pt2'],electron_EB1['eta1'],electron_EB1['eta2'],electron_EB1['phi1'],electron_EB1['phi2'])
m_mix1=minvariant(electron_mix1['pt1'],electron_mix1['pt2'],electron_mix1['eta1'],electron_mix1['eta2'],electron_mix1['phi1'],electron_mix1['phi2'])
fig_5=go.Figure()
fig_5.add_trace(go.Histogram(x=minv1, name='RAW q=0',nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_EB1, name='BARREL q=0',nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_mix1, name='MIXTO q=0',nbinsx=200))
fig_5.add_trace(go.Histogram(x=m_EE1, name='ENDCAP q=0',nbinsx=200))
fig_5.update_layout(barmode='overlay')
fig_5.show()

la longitud de endcap q=0 es  933
la longitud de barrel q=0 es  3893
la longitud de mix q=0 es  2765
la longitud de raw q=0 es  7591


Se van a hacer filtros de tolerancia y se van a descartar datos de ambos electrones en el endcap, la grafica de masa invariante es muy aplanada, ya se hicieron los filtros para los datos raw, se seguiran con mixtos y barrel. Empezaremos limpiando los datos de barrel

In [11]:
tol1=0.011
electron_EBeta=electron_EB1[(electron_EB1.sigmaEtaEta1 <=tol1 )  & (electron_EB1.sigmaEtaEta2 <=tol1)]
m_EBeta=minvariant(electron_EBeta['pt1'],electron_EBeta['pt2'],electron_EBeta['eta1'],electron_EBeta['eta2'],electron_EBeta['phi1'],electron_EBeta['phi2'])
fig_6=go.Figure()
fig_6.add_trace(go.Histogram(x=m_EB, name='eventos EB raw',nbinsx=200))
fig_6.add_trace(go.Histogram(x=m_EB1, name='eventos EB Q=0',nbinsx=200))
fig_6.add_trace(go.Histogram(x=m_EBeta, name='sigmaetaeta<=tol',nbinsx=200))
fig_6.update_layout(barmode='overlay')
fig_6.show()
print("la longitud de los datos del barrel con el filtro Q=0 y sigmaetaeta<0.011 es",len(m_EBeta))

la longitud de los datos del barrel con el filtro Q=0 y sigmaetaeta<0.011 es 2913


In [47]:
tol_iso2=11.5
electron_EBiso=electron_EBeta[(electron_EBeta.isoEcal1+electron_EBeta.isoHcal1+electron_EBeta.isoTrack1<=tol_iso2) 
& (electron_EBeta.isoEcal2+electron_EBeta.isoHcal2+electron_EBeta.isoTrack2<=tol_iso2)]
m_EBiso=minvariant(electron_EBiso['pt1'],electron_EBiso['pt2'],electron_EBiso['eta1'],electron_EBiso['eta2'],electron_EBiso['phi1'],electron_EBiso['phi2'])
fig_7=go.Figure()
fig_7.add_trace(go.Histogram(x=m_EB, name='eventos EB raw',nbinsx=200))
fig_7.add_trace(go.Histogram(x=m_EB1, name='eventos EB Q=0',nbinsx=200))
fig_7.add_trace(go.Histogram(x=m_EBeta, name='sigmaetaeta<=tol',nbinsx=200))
fig_7.add_trace(go.Histogram(x=m_EBiso, name='filtro iso<=tol',nbinsx=200))
fig_7.update_layout(barmode='overlay')
fig_7.show()
print(len(m_EBiso))



2570


Ahora se limpiaran los datos mixtos, aquellos en los que aparecen un EB y un EE 

In [34]:
tol3=0.021
electron_mixeta=electron_mix1[(electron_mix1.sigmaEtaEta1<=tol3) & (electron_mix1.sigmaEtaEta2<=tol3)]
m_mixeta=minvariant(electron_mixeta['pt1'],electron_mixeta['pt2'],electron_mixeta['eta1'],electron_mixeta['eta2'],electron_mixeta['phi1'],electron_mixeta['phi2'])
fig_8=go.Figure()
fig_8.add_trace(go.Histogram(x=m_mix, name='eventos Emix raw',nbinsx=200))
fig_8.add_trace(go.Histogram(x=m_mix1, name='eventos Emix Q=0',nbinsx=200))
fig_8.add_trace(go.Histogram(x=m_mixeta, name='sigmaetaeta<=tol',nbinsx=200))
fig_8.update_layout(barmode='overlay')
fig_8.show()
print("la longitud de los datos del barrel es es",len(m_mix))
print("la longitud de los datos del barrel con el filtro Q=0 es",len(m_mix1))
print("la longitud de los datos del barrel con el filtro Q=0 y sigmaetaeta es",len(m_mixeta))

la longitud de los datos del barrel es es 3747
la longitud de los datos del barrel con el filtro Q=0 es 2765
la longitud de los datos del barrel con el filtro Q=0 y sigmaetaeta es 1841


Se pierden muchos datos alrededor de 91Gev/c^2, no se hará un filtro de sigmaEtaEta, se procede con el filtro iso

In [42]:
tol_iso3=15
electron_mixiso=electron_mix1[(electron_mix1.isoEcal1+electron_mix1.isoHcal1+electron_mix1.isoTrack1<=tol_iso3) 
& (electron_mix1.isoEcal2+electron_mix1.isoHcal2+electron_mix1.isoTrack2<=tol_iso3)]
m_mixiso=minvariant(electron_mixiso['pt1'],electron_mixiso['pt2'],electron_mixiso['eta1'],electron_mixiso['eta2'],electron_mixiso['phi1'],electron_mixiso['phi2'])
fig_9=go.Figure()
fig_9.add_trace(go.Histogram(x=m_mix, name='eventos Emix raw',nbinsx=200))
fig_9.add_trace(go.Histogram(x=m_mix1, name='eventos Emix Q=0',nbinsx=200))
fig_9.add_trace(go.Histogram(x=m_mixiso, name='isotot<=tol',nbinsx=200))
fig_9.update_layout(barmode='overlay')
fig_9.show()
print("la longitud de los datos del barrel es es",len(m_mix))
print("la longitud de los datos del barrel con el filtro Q=0 es",len(m_mix1))
print("la longitud de los datos del barrel con el filtro Q=0 e iso es",len(m_mixiso))

la longitud de los datos del barrel es es 3747
la longitud de los datos del barrel con el filtro Q=0 es 2765
la longitud de los datos del barrel con el filtro Q=0 e iso es 1975


Ahora, comparando las 3 graficas filtradas se obtienen:

In [46]:
fig_10=go.Figure()
fig_10.add_trace(go.Histogram(x=minviso, name='eventos totales filtrados',nbinsx=200))
fig_10.add_trace(go.Histogram(x=m_EBiso, name='eventos Ebarrel filtrados',nbinsx=200))
fig_10.add_trace(go.Histogram(x=m_mixiso, name='eventos Emix filtrados',nbinsx=200))
fig_10.add_trace(go.Histogram(x=m_EE1,name='eventos endcap q=0',nbinsx=200))
fig_10.update_layout(barmode='overlay')
fig_10.show()